format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/DelimitedConfig.java

```diff
@@ -49,4 +66,103 @@
     return skipHeader == null ? false : skipHeader;
   }
+
+  @Nullable
+  @Override
+  public Schema getSchema() {
+    if (containsMacro(NAME_SCHEMA)) {
+      return null;
+    }
+    if (schema == null || schema.equals("")) {
+      try {
+        return getDefaultSchema(null);
+      } catch (IOException e) {
+        throw new IllegalArgumentException("Invalid schema: " + e.getMessage(), e);
+      }
+    }
+    return super.getSchema();
+  }
+
+  /**
+   * Reads delimiter from config
+   * If not available returns default delimiter based on format
+   * @return delimiter
+   */
+  private String getDefaultDelimiter() {
+    String delimiter = getProperties().getProperties().get(DELIMITER);
+    if (delimiter != null) {
+      return delimiter;
+    }
+    final String format = getProperties().getProperties().get(FORMAT);
+    switch (format) {
+      case "tsv":
+        return "\t";
+      default:
+        return ",";
+    }
+  }
+
+  /**
+   * Extract schema from file
+   *
+   * @param context {@link FormatContext}
+   * @return {@link Schema}
+   * @throws IOException raised when error occurs during schema extraction
+   */
+  public Schema getDefaultSchema(FormatContext context) throws IOException {
+    final String format = getProperties().getProperties().getOrDefault(FORMAT, "delimited");
+    String delimiter = getProperties().getProperties().get(DELIMITER);
+    if (format.equals("delimited") && Strings.isNullOrEmpty(delimiter)) {
+      throw new IllegalArgumentException("Delimiter is required when format is set to 'delimited'.");
+    }
+    List<Schema.Field> fields = new ArrayList<>();
+    String path = getProperties().getProperties().getOrDefault(
+      "path", ""
+    );
+
+    Job job = JobUtils.createInstance();
+    Configuration conf = job.getConfiguration();
+    // set entries here, before FileSystem is used
+    for (Map.Entry<String, String> entry : getFileSystemProperties().entrySet()) {
+      conf.set(entry.getKey(), entry.getValue());
+    }
+    FSDataInputStream input = null;
+    BufferedReader bufferedReader = null;
+    String line = null;
+    try {
+      final Path file = getFilePathForSchemaGeneration(path,
+                                                       format.equals("delimited") ? null : format, conf);
+      final FileSystem fileSystem = FileSystem.get(file.toUri(), conf);
+      input = fileSystem.open(file);
+      bufferedReader = new BufferedReader(new InputStreamReader(input));
+      line = bufferedReader.readLine();
+      if (line == null) {
+        return null;
+      }
+    } finally {
+      if (bufferedReader != null) {
+        bufferedReader.close();
+      }
+      if (input != null) {
+        input.close();
+      }
+    }
+    String[] columns = line.split(getDefaultDelimiter());
+    int count = 1;
+    for (String column : columns) {
+      if (getSkipHeader()) {
+        fields.add(Schema.Field.of(column, Schema.of(Schema.Type.STRING)));
```

Comment:
@bdmogal This is Akhilesh. This schema generation goes a long way in auto detecting the column names. But we may have to extend this to infer other types as well.

Reply:
Sure, for CSV, the current requirement was to treat it as all strings, since that's how Wrangler does it today. However, happy to hear thoughts about how you are extending it to other types.

We can track that separately